File: src/main/kotlin/xyz/mahmoudahmed/dsl/TargetPathBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.Function;
import java.util.function.Predicate;

public class TargetPathBuilder {
    private final TargetBuilder parent;
    private final String sourcePath;
    private final String targetPath;
    private Function<Object, Object> transformation;
    private Predicate<Object> condition;

    TargetPathBuilder(TargetBuilder parent, String sourcePath, String targetPath,
                      Function<Object, Object> transformation, Predicate<Object> condition) {
        this.parent = parent;
        this.sourcePath = sourcePath;
        this.targetPath = targetPath;
        this.transformation = transformation;
        this.condition = condition;
    }

    public TargetBuilder using(Function<Object, Object> function) {
        this.transformation = function;
        finalizeMappingRule();
        return parent;
    }

    public TargetBuilder when(Predicate<Object> condition) {
        this.condition = condition;
        finalizeMappingRule();
        return parent;
    }

    // Finalize and return to parent
    public TargetBuilder end() {
        finalizeMappingRule();
        return parent;
    }

    private void finalizeMappingRule() {
        MappingRule rule = new SimpleMapping(sourcePath, targetPath, transformation, condition);
        parent.addRule(rule);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/builders/TargetPathBuilder.kt
package xyz.mahmoudahmed.dsl.builders

import xyz.mahmoudahmed.dsl.core.SimpleMapping

class TargetPathBuilder(
    private val parent: TargetBuilder,
    private val sourcePath: String,
    private val targetPath: String,
    private var transformation: ((Any) -> Any)?,
    private var condition: ((Any) -> Boolean)?
) {
    fun using(function: (Any) -> Any): TargetBuilder {
        this.transformation = function
        finalizeMappingRule()
        return parent
    }

    fun chooseIf(condition: (Any) -> Boolean): TargetBuilder {
        this.condition = condition
        finalizeMappingRule()
        return parent
    }

    // Finalize and return to parent
    fun end(): TargetBuilder {
        finalizeMappingRule()
        return parent
    }

    private fun finalizeMappingRule() {
        val rule = SimpleMapping(sourcePath, targetPath, transformation, condition)
        parent.addRule(rule)
    }
}